src/main/java/org/truffleruby/core/CoreLibrary.java

```diff
@@ -415,7 +415,14 @@ public CoreLibrary(RubyContext context) {
         Layouts.CLASS.setInstanceFactoryUnsafe(fiberClass, fiberFactory);
         defineModule("FileTest");
         hashClass = defineClass("Hash");
-        hashFactory = Layouts.HASH.createHashShape(hashClass, hashClass);
+        final DynamicObjectFactory originalHashFactory = Layouts.HASH.createHashShape(hashClass, hashClass);
+        final DynamicObjectFactory hashFactory;
+        if (context.isPreInitializing()) {
+            hashFactory = context.getPreInitializationManager().hookIntoHashFactory(originalHashFactory);
+        } else {
+            hashFactory = originalHashFactory;
+        }
+        this.hashFactory = hashFactory;
```

Comment:
Why not assign `this.hashFactory` directly in the two branches above?

Reply:
Will fix.